src/db.c

```diff
@@ -332,11 +383,52 @@ robj *dbRandomKey(redisDb *db) {
     }
 }
 
+/* Return random non-empty dictionary from this DB. */
+dict *getFairRandomDict(redisDb *db) {
+    if (db->dict_count == 1) return db->dict[0];
+    unsigned long target = randomULong();
+    unsigned long long int key_count = dbSize(db);
+    if (!key_count) return db->dict[0];
+    target %= key_count; /* Random key index in range [0..KEY_COUNT). */
+    /* In linearly enumerated key space, find a slot that contains target key. */
+    dbIterator dbit;
+    dbInitIterator(&dbit, db);
+    dict *d = NULL;
+    while ((d = dbNextDict(&dbit))) {
```

Comment:
One downside of this is that on large clusters. (~1000 nodes), most nodes are only assigned 16 slots. So they will on average scan through 999 empty slots before finding any slot with keys in it. What's worse, if the cluster is correctly configured and has continuous ranges, they might have to consistently scan 16k empty slots (for example if the node has slots 16358 - 16383). I have to imagine this will take some amount of time. 

I still think we might want to maintain a vector of the dictionaries *owned* by this node we can iterate through. This list will not change often.

Reply:
I think that's a good idea, it would help us speed up iteration and random dict selection at a relatively small cost of <128kb.

Reply:
Can we use the bitmap `myself->slots` for this?

Reply:
It's a smart idea, slot bitmap could speed up iteration, as we can isolate all non zero bits from each byte in the mask quite easily. There are two issues with it though:
1. Similarly to the previous approach, you'd have to skip a bunch of 0s in the bitmap if node owns few slots (although it will be 8 times faster as bitmask array's size is `CLUSTER_SLOTS/8`)
2. It won't be possible to get random non-empty slot in O(1) time using this approach.
So, unfortunately, I don't think it's quite as good as having non empty slots in a vector.

Reply:
A cache line is 64 bytes, so we have 512 bits (slots) in one cold memory read.

Reply:
* Iterating those 512 bits will require some CPU cycles, even if there isn't much random memory access. 
* You may still need multiple reads from memory in order to find a slot if very few slots are owned.
* Problem with finding random non-empty dict isn't solved (it's important that it's O(1) for eviction).

Reply:
I would love to use this bit mask and not have to add/maintain another data structure. However, I fail to see how we can get a random non empty dictionary in O(1) time without additional information of some sort. As Oran said, if node owns few slots (say 1% for example), there will be a lot of guess work involved, unless we have some type of indexed data structure that contains owned slots.